### PR TITLE
fix: Numerical URL targets cause site hang

### PR DESCRIPTION
**Pull request by chenc041**
if location hash is start of numerical, it will cause useScrollRestoration hook throw an error and infinity loop 

change file  [package/remix-react/scroll-restoration.tsx](https://github.com/remix-run/remix/blob/main/packages/remix-react/scroll-restoration.tsx#L101)

issue num [#935](https://github.com/remix-run/remix/issues/935)

reference [Rule sets, declaration blocks, and selectors](https://www.w3.org/TR/CSS2/syndata.html#characters)

### CODE REVIEW

**Review by tascord — COMMENTED** [review]
Looks good!
Not sure about ettiquite in terms of the `.idea` directory in this project so no comment there.
I'll have a look at minifying the scroll-restoration fix or maybe making it a bit more understandable w/ comments.

**Review by tascord — APPROVED** [review]
👍

**Comment by chenc041** [comment]
> 👍

😄

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,4 @@ yarn-error.log
 /fixtures/my-remix-app
 
 .eslintcache
+/.idea/
```
Comment (MichaelDeBoey) [inline .gitignore:14]: This should be in your global `.gitignore` file

https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer

```suggestion
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (BenoitAverty) [reply .gitignore:14]: or if you don't want it global, in .git/info/exclude

File: packages/remix-react/scroll-restoration.tsx [hunk 2]
```diff
@@ -98,7 +98,9 @@ function useScrollRestoration() {
       }
 
       // try to scroll to the hash
-      if (location.hash) {
+      const regular = /^[0-9]\d+/;
+      const { hash } = location;
+      if (hash && !regular.test(hash.replace("#", ""))) {
         let el = document.querySelector(location.hash);
```
Comment (donavon) [inline packages/remix-react/scroll-restoration.tsx:104]: you've destructured `hash` so use it here

```suggestion
        let el = document.querySelector(hash);
```

Comment (chenc041) [reply packages/remix-react/scroll-restoration.tsx]: thank you  so much

**Comment by tascord** [comment]
Hey folk, just wanted to add that some versions of different browsers may allow different characters as ID's, and thus as hashes.
We may just want to wrap things in a try/catch if that works with the convention of the project.
[A little info on id-s](https://mathiasbynens.be/notes/html5-id-class)

**Comment by tascord** [comment]
Or, more specifically :: [here](https://html.spec.whatwg.org/multipage/dom.html#:~:text=There%20are%20no%20other%20restrictions%20on%20what%20form%20an%20ID%20can%20take%3B%20in%20particular%2C%20IDs%20can%20consist%20of%20just%20digits%2C%20start%20with%20a%20digit%2C%20start%20with%20an%20underscore%2C%20consist%20of%20just%20punctuation%2C%20etc.)

**Comment by chenc041** [comment]
> Hey folk, just wanted to add that some versions of different browsers may allow different characters as ID's, and thus as hashes. We may just want to wrap things in a try/catch if that works with the convention of the project. [A little info on id-s](https://mathiasbynens.be/notes/html5-id-class)

if someone use special id-s, i think we give warning in console and give some reference

**Comment by tascord** [comment]
> > Hey folk, just wanted to add that some versions of different browsers may allow different characters as ID's, and thus as hashes. We may just want to wrap things in a try/catch if that works with the convention of the project. [A little info on id-s](https://mathiasbynens.be/notes/html5-id-class)
> 
> if someone use special id-s, i think we give warning in console and give some reference

Sounds good!

**Comment by tascord** [comment]
Not sure how to commit changes myself, but I think this would be a good thing to put in place.
Also not sure if comms are written in the write voicing, so please edit to convention.
```ts
      if (location.hash) {
        if (!/^(#[A-z][A-z0-9-_:.]*)$/.test(location.hash)) {
          return console.warn(
            `The hash supplied, ${location.hash}, is not a valid HTML4 hash (https://www.w3.org/TR/html4/types.html#type-id). As HTML5 hashes aren-t supported in all browsers, scroll restoration will not work with them.`
          );
        }

        let el = document.querySelector(location.hash);
        if (el) {
          el.scrollIntoView();
          return;
        }
      }
```

**Review by tascord — APPROVED** [review]

**Comment by chenc041** [comment]
> Not sure how to commit changes myself, but I think this would be a good thing to put in place. Also not sure if comms are written in the write voicing, so please edit to convention.
> 
> ```ts
>       if (location.hash) {
>         if (!/^(#[A-z][A-z0-9-_:.]*)$/.test(location.hash)) {
>           return console.warn(
>             `The hash supplied, ${location.hash}, is not a valid HTML4 hash (https://www.w3.org/TR/html4/types.html#type-id). As HTML5 hashes aren't supported in all browsers, scroll restoration will not work with them.`
>           );
>         }
> 
>         let el = document.querySelector(location.hash);
>         if (el) {
>           el.scrollIntoView();
>           return;
>         }
>       }
> ```

nice 
I will edit i, thks 🤝

**Review by tascord — APPROVED** [review]

**Comment by ryanflorence** [comment]
Oops, I thought I fixed this already! Thanks for the PR.

- Skip the regex check
- Remove the warning, some people might want numerical hashes for some reason and us warning them all the time would be annoying
- Use `document.getElementById(location.hash.slice(1))` because it doesn't blow up on numerical ids

And then let's get this merged!

**Comment by chenc041** [comment]
> Oops, I thought I fixed this already! Thanks for the PR.
> 
> * Skip the regex check
> * Remove the warning, some people might want numerical hashes for some reason and us warning them all the time would be annoying
> * Use `document.getElementById(location.hash.slice(1))` because it doesn't blow up on numerical ids
> 
> And then let's get this merged!

Hi, I have changed the code follow you suggest

**Review by tascord — APPROVED** [review]

**Comment by dvargas92495** [comment]
Hey, anything blocking this PR? Running into this issue on my end and would love to help get it merged

**Comment by chenc041** [comment]
> Hey, anything blocking this PR? Running into this issue on my end and would love to help get it merged

I think they have forgot...

**Comment by chenc041** [comment]
@ryanflorence hi ryan, If there is no problem, can you merge this PR as soon as possible?

**Comment by na2hiro** [comment]
This fix enables use cases like using hash as URL params `#color=red` and it would be useful. Could you merge this? Thanks!

**Comment by ryanflorence** [comment]
Thanks, sorry been a busy start of the year 😆

**Comment by chenc041** [comment]
> Thanks, sorry been a busy start of the year 😆

thanks, 😄

**Pull request merged**